File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
var mergeTwoLists = function(list1, list2) {
    if (!list1) return list2;
    if (!list2) return list1;
    
    const listsOrderedAsc = (list1.val < list2.val) ? [list1, list2] : [list2, list1];
    
    const newHead = listsOrderedAsc[0];
    newHead.next = mergeTwoLists(listsOrderedAsc[0].next, listsOrderedAsc[1]);
    
    return newHead;
};

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        
        if (list1 == None and list2 == None):
            return None
        elif (list1 == None):
            return list2
        elif (list2 == None):
            return list1
        
        if (list2.val < list1.val):
            return self.mergeTwoLists(list2, list1)
        
        list1.next = self.mergeTwoLists(list1.next, list2)
        return list1